application/comit_node/src/libp2p_bam/handler.rs

```diff
@@ -0,0 +1,383 @@
+    pub response: Response,
+    pub channel: oneshot::Sender<Response>,
+}
+
+#[derive(Debug)]
+pub enum InnerEvent {
+    IncomingRequest(PendingIncomingRequest),
+    IncomingResponse(PendingIncomingResponse),
+    BadIncomingRequest(AutomaticallyGeneratedErrorResponse),
+
+    /// Error variant
+    ///
+    /// This currently covers all errors generated while handling incoming
+    /// frames. Could potentially be expanded to pass more information to
+    /// the NetworkBehaviour.
+    Error,
+}
+
+type BamHandlerEvent = ProtocolsHandlerEvent<BamConfig, (), InnerEvent>;
+
+impl<TSubstream: AsyncRead + AsyncWrite> ProtocolsHandler for BamHandler<TSubstream> {
+    type InEvent = PendingOutgoingRequest;
+    type OutEvent = InnerEvent;
+    type Error = bam::json::Error;
+    type Substream = TSubstream;
+    type InboundProtocol = BamConfig;
+    type OutboundProtocol = BamConfig;
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol> {
+        SubstreamProtocol::new(BamConfig {})
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        protocol: Framed<Negotiated<TSubstream>, JsonFrameCodec>,
+    ) {
+        let current_state = mem::replace(
+            &mut self.network_stream,
+            SubstreamState::Connected(protocol),
+        );
+
+        if let SubstreamState::WaitingForSubstream(task) = current_state {
+            task.notify();
+        }
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        protocol: Framed<Negotiated<TSubstream>, JsonFrameCodec>,
+        _info: Self::OutboundOpenInfo,
+    ) {
+        let current_state = mem::replace(
+            &mut self.network_stream,
+            SubstreamState::Connected(protocol),
+        );
+
+        if let SubstreamState::WaitingForSubstream(task) = current_state {
+            task.notify();
+        }
+    }
+
+    fn inject_event(&mut self, event: Self::InEvent) {
+        let PendingOutgoingRequest { request, channel } = event;
+
+        let frame = request.into_frame(self.next_outgoing_id);
+        self.pending_frames.push_back(frame);
+
+        self.pending_outgoing_request_channels
+            .insert(self.next_outgoing_id, channel);
+
+        self.next_outgoing_id += 1;
+    }
+
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _info: Self::OutboundOpenInfo,
+        _error: ProtocolsHandlerUpgrErr<Infallible>,
+    ) {
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        KeepAlive::Yes
+    }
+
+    fn poll(&mut self) -> Poll<BamHandlerEvent, Self::Error> {
+        self.poll_response_futures();
+
+        match &mut self.network_stream {
+            SubstreamState::Connected(bam_stream) => {
+                futures::try_ready!(bam_stream.poll_complete());
+
+                while let Some(frame) = self.pending_frames.pop_front() {
+                    let message = build_outgoing_log_message(&frame);
```

Comment:
🔧 Just log the the frame here, then below when you actually are able to send it say `"sent frame ${id}" to ${peer}"` or whatever.